File: recursive/modules/verifier/verifier_helper.go
package verifier

import (
	"ExpanderVerifierCircuit/modules/circuit"

	"github.com/consensys/gnark/frontend"
)

func PrepareLayer(
	api frontend.API,
	layer *circuit.Layer,
	alpha frontend.Variable,
	beta frontend.Variable,
	rz0 []frontend.Variable,
	rz1 []frontend.Variable,
	r_simd []frontend.Variable,
	r_mpi []frontend.Variable,
	sp *ScratchPad,
) {
	EqEvalsAtEfficient(
		api,
		rz0,
		alpha,
		sp.EqEvalsAtRz0,
		sp.EqEvalsFirstPart,
		sp.EqEvalsSecondPart,
	)

	if rz1 != nil && beta != nil {
		EqEvalsAtEfficient(
			api,
			rz1,
			beta,
			sp.EqEvalsAtRz1,
			sp.EqEvalsFirstPart,
			sp.EqEvalsSecondPart,
		)

		for i := 0; i < 1<<layer.OutputLenLog; i++ {
			sp.EqEvalsAtRz0[i] = api.Add(sp.EqEvalsAtRz0[i], sp.EqEvalsAtRz1[i])
		}
	}

	EqEvalsAtEfficient(
		api,
		r_simd,
		1,
		sp.EqEvalsAtRSimd,
		sp.EqEvalsFirstPart,
		sp.EqEvalsSecondPart,
	)

	EqEvalsAtEfficient(
		api,
		r_mpi,
		1,
		sp.EqEvalsAtRMpi,
		sp.EqEvalsFirstPart,
		sp.EqEvalsSecondPart,
	)

	sp.RSimd = &r_simd
	sp.RMpi = &r_mpi
}

func EvalCst(
	api frontend.API,
	cst_gates []circuit.Gate,
	public_input [][]frontend.Variable,
	sp *ScratchPad,
) frontend.Variable {
	var v frontend.Variable = 0

	var mpi_size = len(sp.EqEvalsAtRMpi)
	var simd_size = len(sp.EqEvalsAtRSimd)

	if mpi_size != 1 || simd_size != 1 {
		panic("Only support mpi size 1 and simd size 1 for now")
	}

	for i := 0; i < len(cst_gates); i++ {
		var cst_gate circuit.Gate = cst_gates[i]

		var tmp frontend.Variable = 0
		switch cst_gate.Coef.CoefType {
		case circuit.PublicInput:
			n_witnesses := len(public_input)
			if n_witnesses != mpi_size*simd_size {
				panic("Incompatible n_witnesses with mpi and simd size")
			}
			input_idx := cst_gate.Coef.InputIdx
			vals := make([]frontend.Variable, n_witnesses)
			for j := 0; j < n_witnesses; j++ {
				vals[j] = public_input[j][input_idx]
			}

			tmp = CombineWithSimdMpi(api, vals, sp.EqEvalsAtRSimd, sp.EqEvalsAtRMpi)
			tmp = api.Mul(tmp, sp.EqEvalsAtRz0[cst_gate.OId])
		default:
			coef_value := cst_gate.Coef.GetActualLocalValue()
			tmp = api.Mul(sp.EqEvalsAtRz0[cst_gate.OId], coef_value)
		}

		v = api.Add(v, tmp)
	}
	return v
}

func EvalAdd(
	api frontend.API,
	add_gates []circuit.Gate,
	sp *ScratchPad,
) frontend.Variable {
	var v frontend.Variable = 0
	for i := 0; i < len(add_gates); i++ {
		var add_gate = add_gates[i]
		v = api.Add(
			v,
			api.Mul(sp.EqEvalsAtRz0[add_gate.OId], sp.EqEvalsAtRx[add_gate.IIds[0]], add_gate.Coef.GetActualLocalValue()),
		)
	}
	return api.Mul(v, sp.EqRSimdRSimdXY, sp.EqRMpiRMpiXY)
}

func EvalMul(
	api frontend.API,
	mul_gates []circuit.Gate,
	sp *ScratchPad,
) frontend.Variable {
	var v frontend.Variable = 0
	for i := 0; i < len(mul_gates); i++ {
		var mul_gate = mul_gates[i]
		v = api.Add(
			v,
			api.Mul(
				sp.EqEvalsAtRz0[mul_gate.OId],
				sp.EqEvalsAtRx[mul_gate.IIds[0]],
				sp.EqEvalsAtRy[mul_gate.IIds[1]],
				mul_gate.Coef.GetActualLocalValue(),
			),
		)
	}
	return api.Mul(v, sp.EqRSimdRSimdXY, sp.EqRMpiRMpiXY)
}

func SetRx(
	api frontend.API,
	rx []frontend.Variable,
	sp *ScratchPad,
) {
	EqEvalsAtEfficient(
		api,
		rx,
		1,
		sp.EqEvalsAtRx,
		sp.EqEvalsFirstPart,
		sp.EqEvalsSecondPart,
	)
}

func SetRSimdXY(
	api frontend.API,
	r_simd_xy []frontend.Variable,
	sp *ScratchPad,
) {
	sp.EqRSimdRSimdXY = EqVec(api, *sp.RSimd, r_simd_xy)
}

func SetRMPIXY(
	api frontend.API,
	r_mpi_xy []frontend.Variable,
	sp *ScratchPad,
) {
	sp.EqRMpiRMpiXY = EqVec(api, *sp.RMpi, r_mpi_xy)
}

func SetRY(
	api frontend.API,
	r_y []frontend.Variable,
	sp *ScratchPad,
) {
	EqEvalsAtEfficient(
		api,
		r_y,
		1,
		sp.EqEvalsAtRy,
		sp.EqEvalsFirstPart,
		sp.EqEvalsSecondPart,
	)
}

func Degree2Eval(
	api frontend.API,
	p []frontend.Variable, // lag vals at x=0, 1, 2
	x frontend.Variable,
	sp *ScratchPad,
) frontend.Variable {
	var c0 = p[0]
	var c2 = api.Mul(
		sp.Inv2,
		api.Sub(api.Add(p[2], p[0]), p[1], p[1]),
	)
	var c1 = api.Sub(p[1], p[0], c2)
	return api.Add(
		api.Mul(
			api.Add(api.Mul(c2, x), c1),
			x,
		),
		c0,
	)
}

func Degree3Eval(
	api frontend.API,
	p []frontend.Variable, // lag vals at x=0, 1, 2, 3
	x frontend.Variable,
	sp *ScratchPad,
) frontend.Variable {
	return LagEval(api, p, x, sp)
}

func LagEval(
	api frontend.API,
	vals []frontend.Variable,
	x frontend.Variable,
	sp *ScratchPad,
) frontend.Variable {
	if len(vals) != len(sp.Deg3EvalAt) {
		panic("Incorrect length in LagEval")
	}

	var v frontend.Variable = 0
	for i := 0; i < len(vals); i++ {
		var numerator frontend.Variable = 0
		for j := 0; j < len(vals); j++ {
			if j == i {
				continue
			}

			numerator = api.Mul(numerator, api.Sub(x, sp.Deg3EvalAt[j]))
		}
		v = api.Add(v, api.Mul(numerator, sp.Deg3LagDenomsInv[i], vals[i]))
	}
	return v
}

File: sumcheck/src/verifier_helper.rs
use arith::{ExtensionField, Field};
use circuit::{CircuitLayer, CoefType, GateAdd, GateConst, GateMul, GateUni};
use gkr_engine::{ExpanderDualVarChallenge, FieldEngine, FieldType};
use polynomials::EqPolynomial;

use crate::{scratch_pad::VerifierScratchPad, unpack_and_combine};

#[derive(Default)]
pub struct GKRVerifierHelper<F: FieldEngine> {
    phantom: std::marker::PhantomData<F>,
}

impl<F: FieldEngine> GKRVerifierHelper<F> {
    #[allow(clippy::too_many_arguments)]
    #[inline(always)]
    pub fn prepare_layer(
        layer: &CircuitLayer<F>,
        alpha: &Option<F::ChallengeField>,
        challenge: &ExpanderDualVarChallenge<F>,
        sp: &mut VerifierScratchPad<F>,
        is_output_layer: bool,
    ) {
        assert_eq!(alpha.is_none(), challenge.rz_1.is_none());

        if is_output_layer {
            EqPolynomial::<F::ChallengeField>::eq_eval_at(
                &challenge.rz_0,
                &F::ChallengeField::ONE,
                &mut sp.eq_evals_at_rz0,
                &mut sp.eq_evals_first_part,
                &mut sp.eq_evals_second_part,
            );
        } else {
            // use results from previous layer
            let output_len = 1 << challenge.rz_0.len();
            sp.eq_evals_at_rz0[..output_len].copy_from_slice(&sp.eq_evals_at_rx[..output_len]);
            if alpha.is_some() && challenge.rz_1.is_some() {
                let alpha = alpha.unwrap();
                for i in 0..(1usize << layer.output_var_num) {
                    sp.eq_evals_at_rz0[i] += alpha * sp.eq_evals_at_ry[i];
                }
            }
        }

        EqPolynomial::<F::ChallengeField>::eq_eval_at(
            &challenge.r_simd,
            &F::ChallengeField::ONE,
            &mut sp.eq_evals_at_r_simd,
            &mut sp.eq_evals_first_part,
            &mut sp.eq_evals_second_part,
        );

        EqPolynomial::<F::ChallengeField>::eq_eval_at(
            &challenge.r_mpi,
            &F::ChallengeField::ONE,
            &mut sp.eq_evals_at_r_mpi,
            &mut sp.eq_evals_first_part,
            &mut sp.eq_evals_second_part,
        );

        sp.r_simd = &challenge.r_simd;
        sp.r_mpi = &challenge.r_mpi;
    }

    #[inline(always)]
    pub fn eval_cst(
        cst_gates: &[GateConst<F>],
        public_input: &[F::SimdCircuitField],
        sp: &VerifierScratchPad<F>,
    ) -> F::ChallengeField {
        let mut v = F::ChallengeField::zero();

        let mpi_world_size = sp.eq_evals_at_r_mpi.len();
        let local_input_size = public_input.len() / mpi_world_size;

        for cst_gate in cst_gates {
            let tmp = match cst_gate.coef_type {
                CoefType::PublicInput(input_idx) => {
                    let mut input = vec![];
                    for i in 0..mpi_world_size {
                        input.push(public_input[i * local_input_size + input_idx]);
                    }

                    // mpi combined
                    let input_mpi_combined: F::Field = input
                        .iter()
                        .zip(&sp.eq_evals_at_r_mpi)
                        .map(|(v, c)| *c * *v)
                        .sum();

                    // simd combined
                    sp.eq_evals_at_rz0[cst_gate.o_id]
                        * unpack_and_combine::<F::Field>(
                            &input_mpi_combined,
                            &sp.eq_evals_at_r_simd,
                        )
                }
                _ => sp.eq_evals_at_rz0[cst_gate.o_id] * cst_gate.coef,
            };
            v += tmp;
        }

        v
    }

    #[inline(always)]
    pub fn eval_add(add_gates: &[GateAdd<F>], sp: &VerifierScratchPad<F>) -> F::ChallengeField {
        let mut v = F::ChallengeField::zero();
        for add_gate in add_gates {
            v += sp.eq_evals_at_rz0[add_gate.o_id]
                * sp.eq_evals_at_rx[add_gate.i_ids[0]]
                * add_gate.coef;
        }
        v * sp.eq_r_simd_r_simd_xy * sp.eq_r_mpi_r_mpi_xy
    }

    #[inline(always)]
    pub fn eval_mul(mul_gates: &[GateMul<F>], sp: &VerifierScratchPad<F>) -> F::ChallengeField {
        let mut v = F::ChallengeField::zero();
        for mul_gate in mul_gates {
            let tmp = sp.eq_evals_at_rx[mul_gate.i_ids[0]]
                * sp.eq_evals_at_ry[mul_gate.i_ids[1]]
                * mul_gate.coef;
            v += sp.eq_evals_at_rz0[mul_gate.o_id] * tmp;
        }
        v * sp.eq_r_simd_r_simd_xy * sp.eq_r_mpi_r_mpi_xy
    }

    /// GKR2 equivalent of `eval_add`. (Note that GKR2 uses pow1 gates instead of add gates)
    #[inline(always)]
    pub fn eval_pow_1(gates: &[GateUni<F>], sp: &VerifierScratchPad<F>) -> F::ChallengeField {
        let mut v = F::ChallengeField::zero();
        for gate in gates {
            // Gates of type 12346 represent an add gate
            if gate.gate_type == 12346 {
                v += sp.eq_evals_at_rz0[gate.o_id] * sp.eq_evals_at_rx[gate.i_ids[0]] * gate.coef;
            }
        }
        v * sp.eq_r_simd_r_simd_xy * sp.eq_r_mpi_r_mpi_xy
    }

    #[inline(always)]
    pub fn eval_pow_5(gates: &[GateUni<F>], sp: &VerifierScratchPad<F>) -> F::ChallengeField {
        let mut v = F::ChallengeField::zero();
        for gate in gates {
            // Gates of type 12345 represent a pow5 gate
            if gate.gate_type == 12345 {
                v += sp.eq_evals_at_rz0[gate.o_id] * sp.eq_evals_at_rx[gate.i_ids[0]] * gate.coef;
            }
        }
        v * sp.eq_r_simd_r_simd_xy * sp.eq_r_mpi_r_mpi_xy
    }

    #[inline(always)]
    pub fn set_rx(rx: &[F::ChallengeField], sp: &mut VerifierScratchPad<F>) {
        EqPolynomial::<F::ChallengeField>::eq_eval_at(
            rx,
            &F::ChallengeField::ONE,
            &mut sp.eq_evals_at_rx,
            &mut sp.eq_evals_first_part,
            &mut sp.eq_evals_second_part,
        );
    }

    #[inline(always)]
    pub fn set_r_simd_xy(r_simd_xy: &[F::ChallengeField], sp: &mut VerifierScratchPad<F>) {
        sp.eq_r_simd_r_simd_xy = EqPolynomial::<F::ChallengeField>::eq_vec(
            unsafe { sp.r_simd.as_ref().unwrap() },
            r_simd_xy,
        );
    }

    #[inline(always)]
    pub fn set_r_mpi_xy(r_mpi_xy: &[F::ChallengeField], sp: &mut VerifierScratchPad<F>) {
        sp.eq_r_mpi_r_mpi_xy = EqPolynomial::<F::ChallengeField>::eq_vec(
            unsafe { sp.r_mpi.as_ref().unwrap() },
            r_mpi_xy,
        );
    }

    #[inline(always)]
    pub fn set_ry(ry: &[F::ChallengeField], sp: &mut VerifierScratchPad<F>) {
        EqPolynomial::<F::ChallengeField>::eq_eval_at(
            ry,
            &F::ChallengeField::ONE,
            &mut sp.eq_evals_at_ry,
            &mut sp.eq_evals_first_part,
            &mut sp.eq_evals_second_part,
        );
    }

    #[inline(always)]
    pub fn degree_2_eval(
        ps: &[F::ChallengeField],
        x: F::ChallengeField,
        sp: &VerifierScratchPad<F>,
    ) -> F::ChallengeField {
        assert_eq!(ps.len(), 3);
        let p0 = ps[0];
        let p1 = ps[1];
        let p2 = ps[2];

        if F::FIELD_TYPE == FieldType::GF2 {
            let tmp = p0 - p1;

            let c0 = p0;
            let c2 = (p2 - p0 + tmp.mul_by_x()) * sp.gf2_deg2_eval_coef;
            let c1 = tmp - c2;
            c0 + (c2 * x + c1) * x
        } else {
            let c0 = p0;
            let c2 = F::ChallengeField::INV_2 * (p2 - p1 - p1 + p0);
            let c1 = p1 - p0 - c2;
            c0 + (c2 * x + c1) * x
        }
    }

    #[inline(always)]
    pub fn degree_3_eval(
        vals: &[F::ChallengeField],
        x: F::ChallengeField,
        sp: &VerifierScratchPad<F>,
    ) -> F::ChallengeField {
        Self::lag_eval(vals, x, sp)
    }

    #[inline(always)]
    pub fn degree_6_eval(
        vals: &[F::ChallengeField],
        x: F::ChallengeField,
        sp: &VerifierScratchPad<F>,
    ) -> F::ChallengeField {
        Self::lag_eval(vals, x, sp)
    }

    #[inline(always)]
    #[allow(clippy::needless_range_loop)]
    fn lag_eval(
        vals: &[F::ChallengeField],
        x: F::ChallengeField,
        sp: &VerifierScratchPad<F>,
    ) -> F::ChallengeField {
        let (evals, lag_denoms_inv) = match vals.len() {
            4 => (sp.deg3_eval_at.to_vec(), sp.deg3_lag_denoms_inv.to_vec()),
            7 => (sp.deg6_eval_at.to_vec(), sp.deg6_lag_denoms_inv.to_vec()),
            _ => panic!("unsupported degree"),
        };

        let mut v = F::ChallengeField::ZERO;
        for i in 0..vals.len() {
            let mut numerator = F::ChallengeField::ONE;
            for j in 0..vals.len() {
                if j == i {
                    continue;
                }
                numerator *= x - evals[j];
            }
            v += numerator * lag_denoms_inv[i] * vals[i];
        }
        v
    }
}
